devdoc/channel_internal.md

```diff
@@ -0,0 +1,197 @@
+
+**SRS_CHANNEL_INTERNAL_43_092: [** `channel_internal_dispose` shall release the reference to `THANDLE(CHANNEL_INTERNAL)`. **]**
+
+
+### channel_internal_dispose
+```c
+    static void channel_internal_dispose(CHANNEL_INTERNAL* channel_internal);
+```
+
+`channel_internal_dispose` disposes the given `channel_internal`.
+
+**SRS_CHANNEL_INTERNAL_43_099: [** `channel_internal_dispose` shall call `srw_lock_destroy`. **]**
+
+**SRS_CHANNEL_INTERNAL_43_091: [** `channel_internal_dispose` shall release the reference to `THANDLE(THREADPOOL)`. **]**
+
+### channel_internal_pull
+```c
+    MOCKABLE_FUNCTION(, CHANNEL_RESULT, channel_internal_pull, THANDLE(CHANNEL_INTERNAL), channel_internal, PULL_CALLBACK, pull_callback, void*, pull_context, THANDLE(ASYNC_OP)*, out_op_pull);
+```
+
+`channel_internal_pull` registers the given `pull_callback` to be called when there is data to be consumed.
+
+**SRS_CHANNEL_INTERNAL_43_010: [** `channel_internal_pull` shall call `srw_lock_acquire_exclusive`. **]**
+
+**SRS_CHANNEL_INTERNAL_43_101: [** If the list of pending operations is empty or the first operation in the list of pending operations contains a `non-NULL` `pull_callback`: **]**
+
+ - **SRS_CHANNEL_INTERNAL_43_103: [** `channel_internal_pull` shall create a `THANDLE(ASYNC_OP)` by calling `async_op_create` with `cancel_op` as `cancel`. **]**
+
+ - **SRS_CHANNEL_INTERNAL_43_104: [** `channel_internal_pull` shall store the `pull_callback` and `pull_context` in the `THANDLE(ASYNC_OP)`. **]**
+
+ - **SRS_CHANNEL_INTERNAL_43_111: [** `channel_internal_pull` shall set the `result` of the created `operation` to `CHANNEL_CALLBACK_RESULT_OK`. **]**
+
+ - **SRS_CHANNEL_INTERNAL_43_105: [** `channel_internal_pull` shall insert the created `THANDLE(ASYNC_OP)` in the list of pending operations by calling `DList_InsertTailList`. **]**
+
+ - **SRS_CHANNEL_INTERNAL_43_107: [** `channel_internal_pull` shall set `*out_op_pull` to the created `THANDLE(ASYNC_OP)`. **]**
+
+**SRS_CHANNEL_INTERNAL_43_108: [** If the first operation in the list of pending operations contains a `non-NULL` `push_callback`: **]**
+
+ - **SRS_CHANNEL_INTERNAL_43_109: [** `channel_internal_pull` shall call `DList_RemoveHeadList` on the list of pending operations to obtain the `operation`. **]**
+
+ - **SRS_CHANNEL_INTERNAL_43_112: [** `channel_internal_pull` shall store the `pull_callback` and `pull_context` in the obtained `operation`. **]**
+
+ - **SRS_CHANNEL_INTERNAL_43_113: [** `channel_internal_pull` shall call `threadpool_schedule_work` with `execute_callbacks` as `work_function` and the obtained `operation` as `work_function_context`. **]**
+
+ - **SRS_CHANNEL_INTERNAL_43_114: [** `channel_internal_pull` shall set `*out_op_pull` to the `THANDLE(ASYNC_OP)` of the obtained `operation`. **]**
+
+**SRS_CHANNEL_INTERNAL_43_115: [** `channel_internal_pull` shall call `srw_lock_release_exclusive`. **]**
+
```

Comment:
I assume this is the same like before (moved from channel).

Reply:
Yes.